js/ad_manager_vast.js

```diff
@@ -1875,6 +2127,625 @@ OO.Ads.manager(function(_, $) {
+        type: AD_TYPE.INLINE,
+        mediaFile: mediaFile,
+        version: version,
+        durationInMilliseconds: OO.timeStringToSeconds(ad.duration) * 1000
+      };
+
+      return result;
+    }, this);
+
+    /**
+     * Starts the click-to-linear ad
+     * This is only required for VPAID ads
+     * @private
+     * @method Vast#_beginAd
+     */
+    var _beginAd = _.bind(function() {
+      var ad = currentAd.vpaidAd,
+          adLinear = ad.getAdLinear();
+
+      initSkipAdOffset(currentAd);
+      //Since IMA handles its own UI, we want the video player to hide its UI elements
+      this.amc.hidePlayerUi();
+      this.amc.notifyPodStarted(currentAd.id, currentAd.ad.adPodLength);
+
+      this.amc.notifyLinearAdStarted(currentAd.id, {
+        name: currentAd.data.title,
+        duration : ad.getAdDuration(),
+        clickUrl: _hasClickUrl(currentAd),
+        indexInPod: currentAd.ad.sequence,
+        skippable : ad.getAdSkippableState()
+      });
+    }, this);
+
+    /**
+     * Once Ad Playback stopped
+     * This is only required for VPAID ads
+     * @private
+     * @method Vast#_stopAd
+     */
+    var _stopAd = _.bind(function() {
+      if (currentAd && currentAd.vpaidAd) {
+        currentAd.vpaidAd.stopAd();
+      }
+    }, this);
+
+    /**
+     * Gets current ad format, which is either Linear or NonLinear
+     * This is only required for VPAID ads
+     * @private
+     * @method Vast#_getFormat
+     * @return {object} Ad format
+     */
+    var _getFormat = _.bind(function() {
+      var format, name, node;
+      node = this.$_node[0].getElementsByTagName('Linear')[0];
+      if (!node) {
+        node = this.$_node[0].getElementsByTagName('NonLinear')[0];
+      }
+      if (!node) {
+        return;
+      }
+      name = node.nodeName;
+      format = name.toLowerCase() === 'linear' ? 'Linear' : 'NonLinear';
+      return format;
+    }, this);
+
+    /**
+     * Get tracking events.
+     * This is only required for VPAID ads
+     * @public
+     * @method Vast#_getImpressions
+     * @return {array} Array with impressions urls
+     */
+    this._getImpressions = function() {
+      var impressions, node, nodes, _i, _len;
+      impressions = [];
+      nodes = this.$_node[0].getElementsByTagName('Impression');
+      for (_i = 0, _len = nodes.length; _i < _len; _i++) {
+        node = nodes[_i];
+        impressions.push({
+          url: node.textContent
+        });
+      }
+      return impressions;
+    };
+
+    /**
+     * Get tracking events.
+     * This is only required for VPAID ads
+     * @public
+     * @method Vast#_getTracking
+     * @param {object} parent DOM Element to look for tracking events
+     * @return {array} Array with tracking events and urls
+     */
+    this._getTracking = function(parent) {
+      var node, nodes, tracking, _i, _len;
+      tracking = [];
+      nodes = parent.getElementsByTagName('Tracking');
+      if (!nodes) {
+        return;
```

Comment:
Empty array is better but if you can't change it, then specify that it returns undefined in the @return documentation.